<!DOCTYPE HTML>
<html lang="zh-CN">


<head>
    <meta charset="utf-8">
    <meta name="keywords" content="fungiGenomeAssembleAnnotataion, zhangchaofan&#39;s blog">
    <meta name="description" content="分享生活，分享知识">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <meta name="renderer" content="webkit|ie-stand|ie-comp">
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="format-detection" content="telephone=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">  
    
    <style>
        
        #waifu{
            left: auto !important;
            right: 440px !important;
        }
    </style>
    
    <!-- Global site tag (gtag.js) - Google Analytics -->


    <title>fungiGenomeAssembleAnnotataion | zhangchaofan&#39;s blog</title>
    <link rel="icon" type="image/png" href="/favicon.png">
    <link rel="stylesheet" type="text/css" href="/libs/awesome/css/all.css">
    <link rel="stylesheet" type="text/css" href="/libs/materialize/materialize.min.css">
    <link rel="stylesheet" type="text/css" href="/libs/aos/aos.css">
    <link rel="stylesheet" type="text/css" href="/libs/animate/animate.min.css">
    <link rel="stylesheet" type="text/css" href="/libs/lightGallery/css/lightgallery.min.css">
    <link rel="stylesheet" type="text/css" href="/css/matery.css">
    <link rel="stylesheet" type="text/css" href="/css/my.css">

    <script src="/libs/jquery/jquery.min.js"></script>
    <script src="https://sdk.jinrishici.com/v2/browser/jinrishici.js" charset="utf-8"></script>

<meta name="generator" content="Hexo 5.3.0"><link rel="alternate" href="/atom.xml" title="zhangchaofan's blog" type="application/atom+xml">
</head>




<body>
    <header class="navbar-fixed">
        <!-- 元宵节灯笼样式开始 -->
<style>
    @media only screen and (min-width: 1124px) {
        .nav-menu {
            padding-right: 96px;
        }
    }
    @media only screen and (max-width: 760px) {
        .deng-box, .deng-box1 {
            width: 40%;
        }
        .right {
            float: left!important;
        }
    }
    @media only screen and (min-width: 768px) and (max-width: 1024px) {
        .right {
            float: left!important;
        }
    }
    .deng-box {
        position: fixed;
        top: -40px;
        right: -20px;
        z-index: 999;
    }
    .deng-box1 {
        position: fixed;
        top: -30px;
        right: 10px;
        z-index: 999;
    }
    .deng-box1 .deng {
        position: relative;
        width: 120px;
        height: 90px;
        margin: 50px;
        background: #d8000f;
        background: rgba(216, 0, 15, 0.8);
        border-radius: 50% 50%;
        -webkit-transform-origin: 50% -100px;
        -webkit-animation: swing 5s infinite ease-in-out;
        box-shadow: -5px 5px 30px 4px rgba(252, 144, 61, 1);
    }
    .deng {
        position: relative;
        width: 120px;
        height: 90px;
        margin: 50px;
        background: #d8000f;
        background: rgba(216, 0, 15, 0.8);
        border-radius: 50% 50%;
        -webkit-transform-origin: 50% -100px;
        -webkit-animation: swing 3s infinite ease-in-out;
        box-shadow: -5px 5px 50px 4px rgba(250, 108, 0, 1);
    }
    .deng-a {
        width: 100px;
        height: 90px;
        background: #d8000f;
        background: rgba(216, 0, 15, 0.1);
        margin: 12px 8px 8px 10px;
        border-radius: 50% 50%;
        border: 2px solid #dc8f03;
    }
    .deng-b {
        width: 45px;
        height: 90px;
        background: #d8000f;
        background: rgba(216, 0, 15, 0.1);
        margin: -4px 8px 8px 26px;
        border-radius: 50% 50%;
        border: 2px solid #dc8f03;
    }
    .xian {
        position: absolute;
        top: -20px;
        left: 60px;
        width: 2px;
        height: 20px;
        background: #dc8f03;
    }
    .shui-a {
        position: relative;
        width: 5px;
        height: 20px;
        margin: -5px 0 0 59px;
        -webkit-animation: swing 4s infinite ease-in-out;
        -webkit-transform-origin: 50% -45px;
        background: #ffa500;
        border-radius: 0 0 5px 5px;
    }
    .shui-b {
        position: absolute;
        top: 14px;
        left: -2px;
        width: 10px;
        height: 10px;
        background: #dc8f03;
        border-radius: 50%;
    }
    .shui-c {
        position: absolute;
        top: 18px;
        left: -2px;
        width: 10px;
        height: 35px;
        background: #ffa500;
        border-radius: 0 0 0 5px;
    }
    .deng:before {
        position: absolute;
        top: -7px;
        left: 29px;
        height: 12px;
        width: 60px;
        content: " ";
        display: block;
        z-index: 999;
        border-radius: 5px 5px 0 0;
        border: solid 1px #dc8f03;
        background: #ffa500;
        background: linear-gradient(to right, #dc8f03, #ffa500, #dc8f03, #ffa500, #dc8f03);
    }
    .deng:after {
        position: absolute;
        bottom: -7px;
        left: 10px;
        height: 12px;
        width: 60px;
        content: " ";
        display: block;
        margin-left: 20px;
        border-radius: 0 0 5px 5px;
        border: solid 1px #dc8f03;
        background: #ffa500;
        background: linear-gradient(to right, #dc8f03, #ffa500, #dc8f03, #ffa500, #dc8f03);
    }
    @font-face {
        font-family: "华文行楷";
        src: url('/fonts/华文行楷.ttf');
    }
    .deng-t {
        font-family: 华文行楷;
        font-size: 26px;
        color: #dc8f03;
        font-weight: bold;
        line-height: 44px;
        text-align: center;
    }
    .night .deng-t,
    .night .deng-box,
    .night .deng-box1 {
        background: transparent !important;
    }
    @-moz-keyframes swing {
        0% {
                -moz-transform: rotate(-10deg)
        }
        50% {
             -moz-transform: rotate(10deg)
        }
        100% {
                -moz-transform: rotate(-10deg)
        }
    }
    @-webkit-keyframes swing {
        0% {
                -webkit-transform: rotate(-10deg)
        }
        50% {
                -webkit-transform: rotate(10deg)
        }
        100% {
                -webkit-transform: rotate(-10deg)
        }
    }
</style>
<div class="deng-box">
    <div class="deng">
        <div class="xian"></div>
        <div class="deng-a">
            <div class="deng-b"><div class="deng-t">喜迎</div></div>
        </div>
        <div class="shui shui-a"><div class="shui-c"></div><div class="shui-b"></div></div>
    </div>
</div>
<div class="deng-box1">
    <div class="deng">
        <div class="xian"></div>
        <div class="deng-a">
            <div class="deng-b"><div class="deng-t">新春</div></div>
        </div>
        <div class="shui shui-a"><div class="shui-c"></div><div class="shui-b"></div></div>
    </div>
</div>
<!-- 元宵节灯笼样式结束 -->
    <nav id="headNav" class="bg-color nav-transparent">
        <div id="navContainer" class="nav-wrapper container">
            <div class="brand-logo">
                <a href="/" class="waves-effect waves-light">
                    
                    <img src="/favicon.png" class="logo-img" alt="LOGO">
                    
                    <span class="logo-span">zhangchaofan&#39;s blog</span>
                </a>
            </div>
            

<a href="#" data-target="mobile-nav" class="sidenav-trigger button-collapse"><i class="fas fa-bars"></i></a>
<ul class="right nav-menu">
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/" class="waves-effect waves-light">
      
      <i class="fas fa-home" style="zoom: 0.6;"></i>
      
      <span>首页</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/tags" class="waves-effect waves-light">
      
      <i class="fas fa-tags" style="zoom: 0.6;"></i>
      
      <span>标签</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/categories" class="waves-effect waves-light">
      
      <i class="fas fa-bookmark" style="zoom: 0.6;"></i>
      
      <span>分类</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/archives" class="waves-effect waves-light">
      
      <i class="fas fa-archive" style="zoom: 0.6;"></i>
      
      <span>归档</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/about" class="waves-effect waves-light">
      
      <i class="fas fa-user-circle" style="zoom: 0.6;"></i>
      
      <span>关于</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/contact" class="waves-effect waves-light">
      
      <i class="fas fa-comments" style="zoom: 0.6;"></i>
      
      <span>联系我</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="/friends" class="waves-effect waves-light">
      
      <i class="fas fa-address-book" style="zoom: 0.6;"></i>
      
      <span>友联</span>
    </a>
    
  </li>
  
  <li class="hide-on-med-and-down nav-item">
    
    <a href="" class="waves-effect waves-light">

      
      <i class="fas fa-location-arrow" style="zoom: 0.6;"></i>
      
      <span>导航</span>
      <i class="fas fa-chevron-down" aria-hidden="true" style="zoom: 0.6;"></i>
    </a>
    <ul class="sub-nav menus_item_child ">
      
      <li>
        <a href="/navigate">
          
          <i class="fas fa-location-arrow" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>导航</span>
        </a>
      </li>
      
      <li>
        <a href="/home">
          
          <i class="fas fa-home" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>首页</span>
        </a>
      </li>
      
      <li>
        <a href="/music">
          
          <i class="fas fa-music" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>音乐</span>
        </a>
      </li>
      
      <li>
        <a href="/box">
          
          <i class="fas fa-film" style="margin-top: -20px; zoom: 0.6;"></i>
          
          <span>聚宝盒</span>
        </a>
      </li>
      
    </ul>
    
  </li>
  
  <li>
    <a href="#searchModal" class="modal-trigger waves-effect waves-light">
      <i id="searchIcon" class="fas fa-search" title="搜索" style="zoom: 0.85;"></i>
    </a>
  </li>
</ul>


<div id="mobile-nav" class="side-nav sidenav">

    <div class="mobile-head bg-color">
        
        <img src="/favicon.png" class="logo-img circle responsive-img">
        
        <div class="logo-name">zhangchaofan&#39;s blog</div>
        <div class="logo-desc">
            
            分享生活，分享知识
            
        </div>
    </div>

    

    <ul class="menu-list mobile-menu-list">
        
        <li class="m-nav-item">
	  
		<a href="/" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-home"></i>
			
			首页
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/tags" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-tags"></i>
			
			标签
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/categories" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-bookmark"></i>
			
			分类
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/archives" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-archive"></i>
			
			归档
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/about" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-user-circle"></i>
			
			关于
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/contact" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-comments"></i>
			
			联系我
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="/friends" class="waves-effect waves-light">
			
			    <i class="fa-fw fas fa-address-book"></i>
			
			友联
		</a>
          
        </li>
        
        <li class="m-nav-item">
	  
		<a href="javascript:;">
			
				<i class="fa-fw fas fa-location-arrow"></i>
			
			导航
			<span class="m-icon"><i class="fas fa-chevron-right"></i></span>
		</a>
            <ul  style="background:  ;" >
              
                <li>

                  <a href="/navigate " style="margin-left:75px">
				  
				   <i class="fa fas fa-location-arrow" style="position: absolute;left:50px" ></i>
			      
		          <span>导航</span>
                  </a>
                </li>
              
                <li>

                  <a href="/home " style="margin-left:75px">
				  
				   <i class="fa fas fa-home" style="position: absolute;left:50px" ></i>
			      
		          <span>首页</span>
                  </a>
                </li>
              
                <li>

                  <a href="/music " style="margin-left:75px">
				  
				   <i class="fa fas fa-music" style="position: absolute;left:50px" ></i>
			      
		          <span>音乐</span>
                  </a>
                </li>
              
                <li>

                  <a href="/box " style="margin-left:75px">
				  
				   <i class="fa fas fa-film" style="position: absolute;left:50px" ></i>
			      
		          <span>聚宝盒</span>
                  </a>
                </li>
              
            </ul>
          
        </li>
        
        
    </ul>
</div>


        </div>

        
    </nav>

</header>

    



<div class="bg-cover pd-header post-cover" style="background-image: url('/medias/plant/9.jpg')">
    <div class="container" style="right: 0px;left: 0px;">
        <div class="row">
            <div class="col s12 m12 l12">
                <div class="brand">
                    <h1 class="description center-align post-title">fungiGenomeAssembleAnnotataion</h1>
                </div>
            </div>
        </div>
    </div>
</div>




<main class="post-container content">

    
    <link rel="stylesheet" href="/libs/tocbot/tocbot.css">
<style>
    #articleContent h1::before,
    #articleContent h2::before,
    #articleContent h3::before,
    #articleContent h4::before,
    #articleContent h5::before,
    #articleContent h6::before {
        display: block;
        content: " ";
        height: 100px;
        margin-top: -100px;
        visibility: hidden;
    }

    #articleContent :focus {
        outline: none;
    }

    .toc-fixed {
        position: fixed;
        top: 64px;
    }

    .toc-widget {
        width: 345px;
        padding-left: 20px;
    }

    .toc-widget .toc-title {
        padding: 35px 0 15px 17px;
        font-size: 1.5rem;
        font-weight: bold;
        line-height: 1.5rem;
    }

    .toc-widget ol {
        padding: 0;
        list-style: none;
    }

    #toc-content {
        padding-bottom: 30px;
        overflow: auto;
    }

    #toc-content ol {
        padding-left: 10px;
    }

    #toc-content ol li {
        padding-left: 10px;
    }

    #toc-content .toc-link:hover {
        color: #42b983;
        font-weight: 700;
        text-decoration: underline;
    }

    #toc-content .toc-link::before {
        background-color: transparent;
        max-height: 25px;

        position: absolute;
        right: 23.5vw;
        display: block;
    }

    #toc-content .is-active-link {
        color: #42b983;
    }

    #floating-toc-btn {
        position: fixed;
        right: 15px;
        bottom: 76px;
        padding-top: 15px;
        margin-bottom: 0;
        z-index: 998;
    }

    #floating-toc-btn .btn-floating {
        width: 48px;
        height: 48px;
    }

    #floating-toc-btn .btn-floating i {
        line-height: 48px;
        font-size: 1.4rem;
    }
</style>
<div class="row">
    <div id="main-content" class="col s12 m12 l9">
        <!-- 文章内容详情 -->
<div id="artDetail">
    <div class="card">
        <div class="card-content article-info">
            <div class="row tag-cate">
                <div class="col s7">
                    
                    <div class="article-tag">
                        
                            <a href="/tags/Genome/">
                                <span class="chip bg-color">Genome</span>
                            </a>
                        
                            <a href="/tags/Assemble/">
                                <span class="chip bg-color">Assemble</span>
                            </a>
                        
                    </div>
                    
                </div>
                <div class="col s5 right-align">
                    
                </div>
            </div>

            <div class="post-info">
                
                <div class="post-date info-break-policy">
                    <i class="far fa-calendar-minus fa-fw"></i>发布日期:&nbsp;&nbsp;
                    2023-03-26
                </div>
                

                
                <div class="post-date info-break-policy">
                    <i class="far fa-calendar-check fa-fw"></i>更新日期:&nbsp;&nbsp;
                    2023-04-24
                </div>
                

                
                <div class="info-break-policy">
                    <i class="far fa-file-word fa-fw"></i>文章字数:&nbsp;&nbsp;
                    2.8k
                </div>
                

                
                <div class="info-break-policy">
                    <i class="far fa-clock fa-fw"></i>阅读时长:&nbsp;&nbsp;
                    13 分
                </div>
                

            </div>
        </div>
        <hr class="clearfix">

        
        <!-- 是否加载使用自带的 prismjs. -->
        <link rel="stylesheet" href="/libs/prism/prism.css">
        

        

        <div class="card-content article-card-content">
            <div id="articleContent">
                <h2 id="这是一个交链孢霉属物种二代数据基因组组装注释流程"><a href="#这是一个交链孢霉属物种二代数据基因组组装注释流程" class="headerlink" title="这是一个交链孢霉属物种二代数据基因组组装注释流程"></a>这是一个交链孢霉属物种二代数据基因组组装注释流程</h2><h3 id="0-环境准备"><a href="#0-环境准备" class="headerlink" title="0.环境准备"></a>0.环境准备</h3><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">conda create -n genome_assemble
conda activate genome_assemble
conda install fastqc fastuniq bwa kmergenie jellyfish quast -y -c bioconda
# 进入到自己的软件文件夹，没有就新建一个: mkdir software
cd software
wget https:&#x2F;&#x2F;ftp-trace.ncbi.nlm.nih.gov&#x2F;sra&#x2F;sdk&#x2F;3.0.2&#x2F;sratoolkit.2.11.1-centos_linux64.tar.gz
tar -xvzf sratoolkit.2.11.1-centos_linux64.tar.gz
wget http:&#x2F;&#x2F;cab.spbu.ru&#x2F;files&#x2F;release3.15.5&#x2F;SPAdes-3.15.5-Linux.tar.gz
tar -xvzf SPAdes-3.15.5-Linux.tar.gz
wget http:&#x2F;&#x2F;www.usadellab.org&#x2F;cms&#x2F;uploads&#x2F;supplementary&#x2F;Trimmomatic&#x2F;Trimmomatic-0.39.zip
unzip Trimmomatic-0.39.zip

# 下载interproscan
wget https:&#x2F;&#x2F;ftp.ebi.ac.uk&#x2F;pub&#x2F;software&#x2F;unix&#x2F;iprscan&#x2F;5&#x2F;5.61-93.0&#x2F;interproscan-5.61-93.0-64-bit.tar.gz
tar -zxvf interproscan-5.61-93.0-64-bit.tar.gz

# 加入环境变量
export PATH&#x3D;&quot;&#x2F;data&#x2F;chaofan&#x2F;software&#x2F;sratoolkit.2.11.1-centos_linux64&#x2F;bin:$PATH&quot;
export PATH&#x3D;&quot;&#x2F;data&#x2F;chaofan&#x2F;software&#x2F;SPAdes-3.15.5-Linux&#x2F;bin:$PATH&quot;
export PATH&#x3D;&quot;&#x2F;data&#x2F;chaofan&#x2F;software&#x2F;interproscan-5.61-93.0:$PATH&quot;
<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<h3 id="1-下载数据"><a href="#1-下载数据" class="headerlink" title="1.下载数据"></a>1.下载数据</h3><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"># 设置你自己的工作路径，把数据什么的都放在单独的一个工作路径
workdir&#x3D;&#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;01.genome_assemble&#x2F;01.Alternaria_sp
cd $workdir
mkdir 00.raw_data &amp;&amp; cd 00.raw_data
nohup prefetch SRR16888499 -O .&#x2F; &amp;  # RNA-seq
nohup prefetch SRR16914355 -O .&#x2F; &amp;  # WGS(Whole Genome Sequencing)<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<h3 id="2-质控"><a href="#2-质控" class="headerlink" title="2.质控"></a>2.质控</h3><h4 id="2-1-sra2fastq"><a href="#2-1-sra2fastq" class="headerlink" title="2.1 sra2fastq"></a>2.1 sra2fastq</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">fastq-dump --split-3 SRR16888499&#x2F;SRR16888499.sra &amp;
fastq-dump --split-3 SRR16914355&#x2F;SRR16914355.sra &amp;
<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span></span></code></pre>
<h4 id="2-2-raw-fastqc"><a href="#2-2-raw-fastqc" class="headerlink" title="2.2 raw_fastqc"></a>2.2 raw_fastqc</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">fastqc  SRR16888499_1.fastq SRR16888499_2.fastq SRR16914355_1.fastq SRR16914355_2.fastq -t 4 -o .&#x2F;<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>这个是SRR16888499(RNA-seq)的测序质量值和接头报告，我们发现这个质量是非常不错的，但还是使用Trimmomatic对raw_reads进行过滤。<br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppCm9te"><img src="https://s1.ax1x.com/2023/02/27/ppCm9te.png" alt="ppCm9te.png"></a><br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppCmCfH"><img src="https://s1.ax1x.com/2023/02/27/ppCmCfH.png" alt="ppCmCfH.png"></a><br>SRR16914355的质量也很好，我们按同一标准过滤<br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppCmipd"><img src="https://s1.ax1x.com/2023/02/27/ppCmipd.png" alt="ppCmipd.png"></a><br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppCmF1A"><img src="https://s1.ax1x.com/2023/02/27/ppCmF1A.png" alt="ppCmF1A.png"></a></p>
<h4 id="2-3-Trimmomatic-filter"><a href="#2-3-Trimmomatic-filter" class="headerlink" title="2.3 Trimmomatic filter"></a>2.3 Trimmomatic filter</h4><p>注意命令和文件的地址，要换回自己的地址，别傻乎乎的直接用</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"># 将不同的测序引物序列合并为一个fasta文件，这样可以处理不同的测序文件
cat &#x2F;data&#x2F;chaofan&#x2F;software&#x2F;Trimmomatic-0.39&#x2F;adapters&#x2F;*.fa &gt; &#x2F;data&#x2F;chaofan&#x2F;software&#x2F;Trimmomatic-0.39&#x2F;adapters&#x2F;all.fa
# cat完之后检查一下 all.fa 是否是标准的fasta格式，有几个adapter文件末端好像没有换行符，导致cat后 fasta的&gt; + ID的那行接到了上一条序列的序列行了

# 
java -jar &#x2F;data&#x2F;chaofan&#x2F;software&#x2F;Trimmomatic-0.39&#x2F;trimmomatic-0.39.jar PE -phred33 SRR16888499_1.fastq SRR16888499_2.fastq  SRR16888499_1.paired.fastq SRR16888499_1.unpaired.fastq SRR16888499_2.paired.fastq SRR16888499_2.unpaired.fastq ILLUMINACLIP:&#x2F;data&#x2F;pipelines&#x2F;00.RNA_seq&#x2F;Trimmomatic-0.39&#x2F;adapters&#x2F;all.fa:2:30:10:1:TRUE SLIDINGWINDOW:4:20 LEADING:3 TRAILING:3 MINLEN:40 -threads 30
# SRR10283202_1.paired.fastq SRR10283202_2.paired.fastq就是我们TRANSCRIPTOMIC的clean data

# 这里改下名字，不然后面fungap会找不到(只找固定后缀文件)
mv SRR16888499_1.paired.fastq SRR16888499_1.fq
mv SRR16888499_2.paired.fastq SRR16888499_2.fq

java -jar &#x2F;data&#x2F;chaofan&#x2F;software&#x2F;Trimmomatic-0.39&#x2F;trimmomatic-0.39.jar PE -phred33 SRR16914355_1.fastq SRR16914355_2.fastq  SRR16914355_1.paired.fastq SRR16914355_1.unpaired.fastq SRR16914355_2.paired.fastq SRR16914355_2.unpaired.fastq ILLUMINACLIP:&#x2F;data&#x2F;pipelines&#x2F;00.RNA_seq&#x2F;Trimmomatic-0.39&#x2F;adapters&#x2F;all.fa:2:30:10:1:TRUE SLIDINGWINDOW:4:20 LEADING:3 TRAILING:3 MINLEN:40 -threads 30
# SRR22980490_1.paired.fastq SRR22980490_2.paired.fastq就是我们WGS的clean data,但还要去一下重复序列

# 参数的具体含义建议去官网上搜索，这里不详细讲述了，参数选择并不唯一<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<h4 id="2-4-remove-PCR-duplication"><a href="#2-4-remove-PCR-duplication" class="headerlink" title="2.4 remove PCR duplication"></a>2.4 remove PCR duplication</h4><p>在基因组组装前，需要对WGS的测序数据去除PCR重复，这里使用fastuniq操作。虽然raw_data就已经没什么重复序列了，但我们还是去一遍，转录组数据一般不用去除这些重复序列。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">mkdir -p ..&#x2F;01.clean_data &amp;&amp; mv *.paired.* ..&#x2F;01.clean_data
cd ..&#x2F;01.clean_data 
# Remove exact duplicates
ls SRR16914355_1.paired.fastq SRR16914355_2.paired.fastq &gt; sample_lst
fastuniq -i sample_lst -o SRR16914355_1.rd.fastq -p SRR16914355_2.rd.fastq <span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<h4 id="2-5-clean-fastqc"><a href="#2-5-clean-fastqc" class="headerlink" title="2.5 clean_fastqc"></a>2.5 clean_fastqc</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">fastqc SRR16914355_1.rd.fastq SRR16914355_2.rd.fastq SRR16888499_1.fq SRR16888499_2.fq -t 4 -o.&#x2F;<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>&emsp;&emsp;这里就不放图了，不知道上传的是不是就是clean_data，质量很好。</p>
<h3 id="3-基因组组装"><a href="#3-基因组组装" class="headerlink" title="3.基因组组装"></a>3.基因组组装</h3><h4 id="3-1-genome-survey"><a href="#3-1-genome-survey" class="headerlink" title="3.1 genome survey"></a>3.1 genome survey</h4><p>kmer在生物信息学上有非常广泛的运用，有非常多方法都是基于kmer的，感兴趣的自己去了解下吧。这里主要使用kmergenie来分析。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">ls SRR16914355_1.rd.fastq SRR16914355_2.rd.fastq &gt; kmer_lst
kmergenie kmer_lst -o kmergenie_result&#x2F;first_kmer -k 127 -l 27 -s 6 -t 40<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span></span></code></pre>
<p><strong>kmergenie log</strong></p>
<pre class="line-numbers language-python" data-language="python"><code class="language-python">running histogram estimation
File sample_lst starts <span class="token keyword">with</span> character <span class="token string">"S"</span><span class="token punctuation">,</span> hence <span class="token keyword">is</span> interpreted <span class="token keyword">as</span> a <span class="token builtin">list</span> of <span class="token builtin">file</span> names
Reading <span class="token number">2</span> read files
Linear estimation<span class="token punctuation">:</span> <span class="token operator">~</span><span class="token number">1910</span> M distinct <span class="token number">74</span><span class="token operator">-</span>mers are <span class="token keyword">in</span> the reads
K<span class="token operator">-</span>mer sampling<span class="token punctuation">:</span> <span class="token number">1</span><span class="token operator">/</span><span class="token number">364</span>
<span class="token operator">|</span> processing  <span class="token operator">|</span>
<span class="token punctuation">[</span>going to estimate histograms <span class="token keyword">for</span> values of k<span class="token punctuation">:</span> <span class="token number">127</span> <span class="token number">121</span> <span class="token number">115</span> <span class="token number">109</span> <span class="token number">103</span> <span class="token number">97</span> <span class="token number">91</span> <span class="token number">85</span> <span class="token number">79</span> <span class="token number">73</span> <span class="token number">67</span> <span class="token number">61</span> <span class="token number">55</span> <span class="token number">49</span> <span class="token number">43</span> <span class="token number">37</span> <span class="token number">31</span>
<span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span>
<span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span>Total time Wallclock  <span class="token number">1924.2</span> s
fitting model to histograms to estimate best k
estimation of the best k so far<span class="token punctuation">:</span> <span class="token number">109</span>
refining estimation around <span class="token punctuation">[</span><span class="token number">103</span><span class="token punctuation">;</span> <span class="token number">115</span><span class="token punctuation">]</span><span class="token punctuation">,</span> <span class="token keyword">with</span> a step of <span class="token number">2</span>
running histogram estimation
File sample_lst starts <span class="token keyword">with</span> character <span class="token string">"S"</span><span class="token punctuation">,</span> hence <span class="token keyword">is</span> interpreted <span class="token keyword">as</span> a <span class="token builtin">list</span> of <span class="token builtin">file</span> names
Reading <span class="token number">2</span> read files
Linear estimation<span class="token punctuation">:</span> <span class="token operator">~</span><span class="token number">2066</span> M distinct <span class="token number">68</span><span class="token operator">-</span>mers are <span class="token keyword">in</span> the reads
K<span class="token operator">-</span>mer sampling<span class="token punctuation">:</span> <span class="token number">1</span><span class="token operator">/</span><span class="token number">394</span>
<span class="token operator">|</span> processing  <span class="token operator">|</span>
<span class="token punctuation">[</span>going to estimate histograms <span class="token keyword">for</span> values of k<span class="token punctuation">:</span> <span class="token number">115</span> <span class="token number">113</span> <span class="token number">111</span> <span class="token number">109</span> <span class="token number">107</span> <span class="token number">105</span> <span class="token number">103</span>
<span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span><span class="token operator">-</span>Total time Wallclock  <span class="token number">798.288</span> s
fitting model to histograms to estimate best k
table of predicted num<span class="token punctuation">.</span> of genomic k<span class="token operator">-</span>mers<span class="token punctuation">:</span> kmergenie_result<span class="token punctuation">.</span>dat
recommended coverage cut<span class="token operator">-</span>off <span class="token keyword">for</span> best k<span class="token punctuation">:</span> <span class="token number">1</span>
best k<span class="token punctuation">:</span> <span class="token number">111</span><span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>最佳<strong>K=111</strong>，我们再跑一遍kmergenie，这次缩小下范围</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">kmergenie kmer_lst -o kmergenie_result&#x2F;re_kmer -k 115 -l 107 -s 2 -t 40<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>这次最佳还是<strong>K=111</strong>，后续我们使用SPAdes组装也要添加这个k值。预估的基因组大约<strong>72,607,905bp</strong>。</p>
<p><a target="_blank" rel="noopener" href="https://imgse.com/i/ppC2ItP"><img src="https://s1.ax1x.com/2023/02/28/ppC2ItP.png" alt="ppC2ItP.png"></a><br>这里是NCBI上公布的基因组序列信息，大小和预估出来的差不太多。<br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppfFWgx"><img src="https://s1.ax1x.com/2023/04/02/ppfFWgx.png" alt="ppfFWgx.png"></a></p>
<h4 id="3-2-SPAdes-assemble"><a href="#3-2-SPAdes-assemble" class="headerlink" title="3.2 SPAdes assemble"></a>3.2 SPAdes assemble</h4><p>主要使用<a target="_blank" rel="noopener" href="https://github.com/ablab/spades">spades</a>来进行基因组组装。软件的具体参数还是建议去GitHub上自己看，直接用这个也行。spades要求kmer长度必须为奇数。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">nohup &#x2F;data&#x2F;chaofan&#x2F;software&#x2F;SPAdes-3.15.5-Linux&#x2F;bin&#x2F;spades.py -k 21,33,55,77,99,109,111,127 -t 20 --careful --cov-cutoff auto -1 SRR16914355_1.rd.fastq -2 SRR16914355_2.rd.fastq -o ..&#x2F;02.draft_genome &amp;<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<h4 id="3-3-quast评估基因组组装质量"><a href="#3-3-quast评估基因组组装质量" class="headerlink" title="3.3 quast评估基因组组装质量"></a>3.3 quast评估基因组组装质量</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">cd ..&#x2F;02.draft_genome
conda activate genome_assemble
&#x2F;data&#x2F;chaofan&#x2F;miniconda&#x2F;envs&#x2F;genome_assemble&#x2F;bin&#x2F;quast scaffolds.fasta
conda deactivate<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span></span></code></pre>
<p>这个是单纯的二代测序数据的组装结果，N50 126.245kbp。<br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppiZUVf"><img src="https://s1.ax1x.com/2023/03/01/ppiZUVf.png" alt="ppiZUVf.png"></a></p>
<h4 id="3-4-BUSCO评估组装一致性"><a href="#3-4-BUSCO评估组装一致性" class="headerlink" title="3.4 BUSCO评估组装一致性"></a>3.4 BUSCO评估组装一致性</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"># 单独新建一个环境配置
# busco 5.4.4有一个BUG https:&#x2F;&#x2F;gitlab.com&#x2F;ezlab&#x2F;busco&#x2F;-&#x2F;issues&#x2F;627
# 很遗憾 busco 5.4.5运行这个DB:dothideomycetes_odb10也会报错(emmm 只有在对基因组使用这个DB评估的时候才有问题,注释蛋白的评估是没问题的) 请安装较老一点版本的busco(5.2.*)。 这里我们直接换一个DB吧 

conda create -n busco.5.4.5 -c conda-forge -c bioconda busco&#x3D;5.4.5
conda activate busco.5.4.5
busco -c 8 -i scaffolds.fasta -l &#x2F;data&#x2F;chaofan&#x2F;source&#x2F;busco_db&#x2F;fungi_odb10 -m geno -o busco.5.4.5_fungi_odb10
conda deactivate<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>

<p><a target="_blank" rel="noopener" href="https://imgse.com/i/ppk19IA"><img src="https://s1.ax1x.com/2023/03/03/ppk19IA.png" alt="ppk19IA.png"></a></p>
<h3 id="4-基因注释"><a href="#4-基因注释" class="headerlink" title="4.基因注释"></a>4.基因注释</h3><p>基因注释主要使用<a target="_blank" rel="noopener" href="https://github.com/CompSynBioLab-KoreaUniv/FunGAP">FunGAP</a>来做，如果不通过docker安装的话，在使用过程中会遇到各种各样的错误，善用google|bing，一般都是可以解决的(<font color=Aqua>这应该是整个过程中最折腾的</font>)。我这里新建了一个conda环境用来基因注释，所以这里我们先切换下环境：</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">conda activate fungap<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>FunGAP的安装请遵循<a target="_blank" rel="noopener" href="https://github.com/CompSynBioLab-KoreaUniv/FunGAP/blob/master/INSTALL.md">Github</a>，这里就不展开讲了</p>
<p><strong>基因注释主要需要三个输入</strong>:</p>
<ol>
<li>组装好的基因组文件;</li>
<li>近源物种的蛋白序列;</li>
<li>August预测 物种模型的选择;</li>
<li>最好测个转录组, FunGAP强制必须加上RNA-seq数据，无论是clean_data还是比对后的bam文件(这个可以修改，但是大部分的脚本文件都需要相应的调整，也可以从NCBI上下载RNA-seq数据)。</li>
</ol>
<h4 id="4-1-获取近源物种的蛋白"><a href="#4-1-获取近源物种的蛋白" class="headerlink" title="4.1 获取近源物种的蛋白"></a>4.1 获取近源物种的蛋白</h4><p>FunGAP提供了脚本，可以直接从ncbi下载近源物种的蛋白序列。你可以在你的FunGAP文件夹下面找到<code>download_sister_orgs.py</code>，这里我们通过全局路径调用。如果可以的话，尽量下载同属或者同科的所有蛋白序列。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">cd $workdir
mkdir 03.gene_anno &amp;&amp; cd 03.gene_anno
python &#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;00.FunGAP&#x2F;FunGAP&#x2F;download_sister_orgs.py \
    --download_dir sister_orgs \
    --taxon &quot;Alternaria&quot; \
    --num_sisters 4 \
    --email_address 2439555626@qq.com 

zcat sister_orgs&#x2F;*faa.gz &gt; prot_db.faa<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<ul>
<li>–taxon 物种名或属名 它会通过ncbi的Taxonomy遍历近源物种的Assembly_data去下载已公布基因组的蛋白数据;</li>
<li>–num_sisters 下载的近源物种的数量,一般来说越多越好;</li>
<li>–email_address 邮箱地址，随便填.</li>
</ul>
<h4 id="4-2-挑选August预测的物种模型"><a href="#4-2-挑选August预测的物种模型" class="headerlink" title="4.2 挑选August预测的物种模型"></a>4.2 挑选August预测的物种模型</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">python &#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;00.FunGAP&#x2F;FunGAP&#x2F;get_augustus_species.py \
    --genus_name &quot;Alternaria&quot; \
    --email_address 2439555626@qq.com 

# --genus_name 物种名或属名 依然通过ncbi的Taxonomy找到 
# 已有的近源物种的augustus模型，可以直接用<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<pre class="line-numbers language-none"><code class="language-none">Suggestions:
We found them in subphylum level: Pezizomycotina
aspergillus_fumigatus
aspergillus_nidulans
aspergillus_oryzae
aspergillus_terreus
botrytis_cinerea
chaetomium_globosum
coccidioides_immitis
fusarium_graminearum
histoplasma_capsulatum
magnaporthe_grisea
neurospora_crassa<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>这里我就简单的选了第一个直接用了。</p>
<h4 id="4-3-FunGAP进行基因注释"><a href="#4-3-FunGAP进行基因注释" class="headerlink" title="4.3 FunGAP进行基因注释"></a>4.3 FunGAP进行基因注释</h4><p> emmmmmm，有一个隐藏的问题，<code>maker train</code>的过程中会用到TRF对序列进行处理，<code>TRF</code>应该设置了一个256个字符长度的数组来储存输入文件的全局路径，如果你的输入文件路径大于256，TRF就会<code>buffer overflow detected</code>，老倒霉蛋了。这个问题好像还是第一次发现的，特别是通过mpirun并行maker，会给输入文件加上一堆特殊含义的后缀什么的(这也和我改了maker的默认TMPDIR有关，我把它设置为当前运行路径下，可能当前全局路径本身就很长了)。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">&#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;00.FunGAP&#x2F;FunGAP&#x2F;fungap.py \
  --genome_assembly ..&#x2F;02.draft_genome&#x2F;scaffolds.fasta \
  --trans_read_1 SRR22973808_1.fastq \
  --trans_read_2 SRR22973808_2.fastq \
  --augustus_species aspergillus_fumigatus \
  --busco_dataset &#x2F;data&#x2F;chaofan&#x2F;source&#x2F;busco_db&#x2F;dothideomycetes_odb10 \
  --sister_proteome prot_db.faa \
  --num_cores 32 \
  --output_dir Alternaria<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<h4 id="4-4-BUSCO评估基因注释结果-这里用的dothideomycetes-odb10"><a href="#4-4-BUSCO评估基因注释结果-这里用的dothideomycetes-odb10" class="headerlink" title="4.4 BUSCO评估基因注释结果(这里用的dothideomycetes_odb10)"></a>4.4 BUSCO评估基因注释结果(这里用的dothideomycetes_odb10)</h4><pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"># 检查使用二代数据组装基因组基因注释的busco值
busco -c 8 -i Alternaria&#x2F;fungap_out&#x2F;fungap_out_prot.faa -l &#x2F;data&#x2F;chaofan&#x2F;source&#x2F;busco_db&#x2F;dothideomycetes_odb10 -m prot -o Alternaria_dothideomycetes_odb10<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span></span></code></pre>
<p>要是植物有这么高就好了。增加prot_db.faa里面同源物种的蛋白数量会适当的提高busco值，可能会加个百分之零点几吧。<br><a target="_blank" rel="noopener" href="https://imgse.com/i/ppMRuGV"><img src="https://s1.ax1x.com/2023/03/12/ppMRuGV.png" alt="ppMRuGV.png"></a></p>
<h4 id="4-5-检查gff文件"><a href="#4-5-检查gff文件" class="headerlink" title="4.5 检查gff文件"></a>4.5 检查gff文件</h4><p>前段时间被gff文件中的CDS phase搞了一波，以后如果我做完基因注释，绝对检查一遍。使用<a target="_blank" rel="noopener" href="https://github.com/NAL-i5K/GFF3toolkit">GFF3toolkit
</a></p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">gff3_QC -g ..&#x2F;03.gene_anno&#x2F;Alternaria&#x2F;fungap_out&#x2F;fungap_out.gff3 -f scaffolds.fasta -o error.txt -s statistic.txt<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>我们这里可能有一个小问题:</p>
<pre class="line-numbers language-python" data-language="python"><code class="language-python"><span class="token punctuation">[</span><span class="token string">'Line 3'</span><span class="token punctuation">]</span>      Esf0003 Error   <span class="token punctuation">[</span>Strand information missing<span class="token punctuation">:</span> legal chacracter<span class="token punctuation">,</span> <span class="token string">"."</span><span class="token punctuation">,</span> found at the strand field<span class="token punctuation">]</span><span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<p>这个是由于我们的gff文件中还包含了序列长度信息行，这个可以不用管。假设你的gff确实有别的问题(注意看error.txt)，则用gff3_fix进行矫正，gff3_fix也是属于<code>GFF3toolkit</code>的一个工具</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell">gff3_fix -qc_r error.txt -g ..&#x2F;03.gene_anno&#x2F;Alternaria&#x2F;fungap_out&#x2F;fungap_out.gff3 -og ..&#x2F;03.gene_anno&#x2F;Alternaria&#x2F;fungap_out&#x2F;corrected.gff3<span aria-hidden="true" class="line-numbers-rows"><span></span></span></code></pre>
<h4 id="4-6-对注释蛋白进行功能注释"><a href="#4-6-对注释蛋白进行功能注释" class="headerlink" title="4.6 对注释蛋白进行功能注释"></a>4.6 对注释蛋白进行功能注释</h4><p>这里我们使用interproscan.sh来对注释的蛋白进行功能注释，并添加到gff文件中。</p>
<pre class="line-numbers language-shell" data-language="shell"><code class="language-shell"># 删除序列末尾的 * 号
sed -i  &#39;s&#x2F;\*$&#x2F;&#x2F;g&#39; fungap_out_prot.faa

&#x2F;data&#x2F;chaofan&#x2F;software&#x2F;interproscan-5.61-93.0&#x2F;interproscan.sh -i fungap_out_prot.faa -f tsv -appl Pfam --goterms -pa --iprlookup -b  Alternaria --tempdir &#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;maker_tmp&#x2F;

# 将interproscan注释加到gff文件内
&#x2F;data&#x2F;pipelines&#x2F;01.Gene_anno&#x2F;00.FunGAP&#x2F;FunGAP&#x2F;gff3_add_pfam.py --input_gff3 fungap_out.gff3 --pfam_file Alternaria.tsv

# 默认注释是加在mRNA 第一个CDS行的第9列(key 列)，你也可以直接修改这个脚本，把他加在mRNA的后面。
<span aria-hidden="true" class="line-numbers-rows"><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></span></code></pre>
<p>如有任何问题，恳请联系我修正。</p>

                
            </div>
            <hr/>

            

    <div class="reprint" id="reprint-statement">
        
            <div class="reprint__author">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-user">
                        文章作者:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="/about" rel="external nofollow noreferrer">zhangchaofan</a>
                </span>
            </div>
            <div class="reprint__type">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-link">
                        文章链接:
                    </i>
                </span>
                <span class="reprint-info">
                    <a href="https://zhangchaofan01.gitee.io/2023/03/26/fungigenomeassembleannotation/">https://zhangchaofan01.gitee.io/2023/03/26/fungigenomeassembleannotation/</a>
                </span>
            </div>
            <div class="reprint__notice">
                <span class="reprint-meta" style="font-weight: bold;">
                    <i class="fas fa-copyright">
                        版权声明:
                    </i>
                </span>
                <span class="reprint-info">
                    本博客所有文章除特別声明外，均采用
                    <a href="https://creativecommons.org/licenses/by/4.0/deed.zh" rel="external nofollow noreferrer" target="_blank">CC BY 4.0</a>
                    许可协议。转载请注明来源
                    <a href="/about" target="_blank">zhangchaofan</a>
                    !
                </span>
            </div>
        
    </div>

    <script async defer>
      document.addEventListener("copy", function (e) {
        let toastHTML = '<span>复制成功，请遵循本文的转载规则</span><button class="btn-flat toast-action" onclick="navToReprintStatement()" style="font-size: smaller">查看</a>';
        M.toast({html: toastHTML})
      });

      function navToReprintStatement() {
        $("html, body").animate({scrollTop: $("#reprint-statement").offset().top - 80}, 800);
      }
    </script>



            <div class="tag_share" style="display: block;">
                <div class="post-meta__tag-list" style="display: inline-block;">
                    
                        <div class="article-tag">
                            
                                <a href="/tags/Genome/">
                                    <span class="chip bg-color">Genome</span>
                                </a>
                            
                                <a href="/tags/Assemble/">
                                    <span class="chip bg-color">Assemble</span>
                                </a>
                            
                        </div>
                    
                </div>
                <div class="post_share" style="zoom: 80%; width: fit-content; display: inline-block; float: right; margin: -0.15rem 0;">
                    <link rel="stylesheet" type="text/css" href="/libs/share/css/share.min.css">
<div id="article-share">

    
    <div class="social-share" data-sites="twitter,facebook,google,qq,qzone,wechat,weibo,douban,linkedin" data-wechat-qrcode-helper="<p>微信扫一扫即可分享！</p>"></div>
    <script src="/libs/share/js/social-share.min.js"></script>
    

    

</div>

                </div>
            </div>
            
        </div>
    </div>

    

    

    

    

    
        <style>
    .valine-card {
        margin: 1.5rem auto;
    }

    .valine-card .card-content {
        padding: 20px 20px 5px 20px;
    }

    #vcomments textarea {
        box-sizing: border-box;
        background: url("/medias/comment_bg1.webp") 100% 100% no-repeat;
    }

    #vcomments p {
        margin: 2px 2px 10px;
        font-size: 1.05rem;
        line-height: 1.78rem;
    }

    #vcomments blockquote p {
        text-indent: 0.2rem;
    }

    #vcomments a {
        padding: 0 2px;
        color: #4cbf30;
        font-weight: 500;
        text-decoration: none;
    }

    #vcomments img {
        max-width: 100%;
        height: auto;
        cursor: pointer;
    }

    #vcomments ol li {
        list-style-type: decimal;
    }

    #vcomments ol,
    ul {
        display: block;
        padding-left: 2em;
        word-spacing: 0.05rem;
    }

    #vcomments ul li,
    ol li {
        display: list-item;
        line-height: 1.8rem;
        font-size: 1rem;
    }

    #vcomments ul li {
        list-style-type: disc;
    }

    #vcomments ul ul li {
        list-style-type: circle;
    }

    #vcomments table, th, td {
        padding: 12px 13px;
        border: 1px solid #dfe2e5;
    }

    #vcomments table, th, td {
        border: 0;
    }

    table tr:nth-child(2n), thead {
        background-color: #fafafa;
    }

    #vcomments table th {
        background-color: #f2f2f2;
        min-width: 80px;
    }

    #vcomments table td {
        min-width: 80px;
    }

    #vcomments h1 {
        font-size: 1.85rem;
        font-weight: bold;
        line-height: 2.2rem;
    }

    #vcomments h2 {
        font-size: 1.65rem;
        font-weight: bold;
        line-height: 1.9rem;
    }

    #vcomments h3 {
        font-size: 1.45rem;
        font-weight: bold;
        line-height: 1.7rem;
    }

    #vcomments h4 {
        font-size: 1.25rem;
        font-weight: bold;
        line-height: 1.5rem;
    }

    #vcomments h5 {
        font-size: 1.1rem;
        font-weight: bold;
        line-height: 1.4rem;
    }

    #vcomments h6 {
        font-size: 1rem;
        line-height: 1.3rem;
    }

    #vcomments p {
        font-size: 1rem;
        line-height: 1.5rem;
    }

    #vcomments hr {
        margin: 12px 0;
        border: 0;
        border-top: 1px solid #ccc;
    }

    #vcomments blockquote {
        margin: 15px 0;
        border-left: 5px solid #42b983;
        padding: 1rem 0.8rem 0.3rem 0.8rem;
        color: #666;
        background-color: rgba(66, 185, 131, .1);
    }

    #vcomments pre {
        font-family: monospace, monospace;
        padding: 1.2em;
        margin: .5em 0;
        background: #272822;
        overflow: auto;
        border-radius: 0.3em;
        tab-size: 4;
    }

    #vcomments code {
        font-family: monospace, monospace;
        padding: 1px 3px;
        font-size: 0.92rem;
        color: #e96900;
        background-color: #f8f8f8;
        border-radius: 2px;
    }

    #vcomments pre code {
        font-family: monospace, monospace;
        padding: 0;
        color: #e8eaf6;
        background-color: #272822;
    }

    #vcomments pre[class*="language-"] {
        padding: 1.2em;
        margin: .5em 0;
    }

    #vcomments code[class*="language-"],
    pre[class*="language-"] {
        color: #e8eaf6;
    }

    #vcomments [type="checkbox"]:not(:checked), [type="checkbox"]:checked {
        position: inherit;
        margin-left: -1.3rem;
        margin-right: 0.4rem;
        margin-top: -1px;
        vertical-align: middle;
        left: unset;
        visibility: visible;
    }

    #vcomments b,
    strong {
        font-weight: bold;
    }

    #vcomments dfn {
        font-style: italic;
    }

    #vcomments small {
        font-size: 85%;
    }

    #vcomments cite {
        font-style: normal;
    }

    #vcomments mark {
        background-color: #fcf8e3;
        padding: .2em;
    }

    #vcomments table, th, td {
        padding: 12px 13px;
        border: 1px solid #dfe2e5;
    }

    table tr:nth-child(2n), thead {
        background-color: #fafafa;
    }

    #vcomments table th {
        background-color: #f2f2f2;
        min-width: 80px;
    }

    #vcomments table td {
        min-width: 80px;
    }

    #vcomments [type="checkbox"]:not(:checked), [type="checkbox"]:checked {
        position: inherit;
        margin-left: -1.3rem;
        margin-right: 0.4rem;
        margin-top: -1px;
        vertical-align: middle;
        left: unset;
        visibility: visible;
    }
</style>

<div class="card valine-card" data-aos="fade-up">
    <div class="comment_headling" style="font-size: 20px; font-weight: 700; position: relative; padding-left: 20px; top: 15px; padding-bottom: 5px;">
        <i class="fas fa-comments fa-fw" aria-hidden="true"></i>
        <span>评论</span>
    </div>
    <div id="vcomments" class="card-content" style="display: grid">
    </div>
</div>

<script src="/libs/valine/av-min.js"></script>
<script src="/libs/valine/Valine.min.js"></script>
<script>
    new Valine({
        el: '#vcomments',
        appId: 'QHvKVyB4LA2Ds9UxY8p2cCBv-gzGzoHsz',
        appKey: '4eMYY2viJug47FfkfPgz0HBG',
        notify: 'true' === 'true',
        verify: 'true' === 'true',
        visitor: 'true' === 'true',
        avatar: 'identicon',
        pageSize: '10',
        lang: 'zh-cn',
        placeholder: '欢迎指点江山😀，留下你的评论，我一直都在哦！'
    });
</script>

    

    

    

    

<article id="prenext-posts" class="prev-next articles">
    <div class="row article-row">
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge left-badge text-color">
                <i class="fas fa-chevron-left"></i>&nbsp;上一篇</div>
            <div class="card">
                <a href="/2023/04/15/newton-iterative/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/plant/3.webp" class="responsive-img" alt="Newton_iterative">
                        
                        <span class="card-title">Newton_iterative</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                        <span class="publish-date">
                            <i class="far fa-clock fa-fw icon-date"></i>2023-04-15
                        </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            zhangchaofan
                            
                        </span>
                    </div>
                </div>
                
                <div class="card-action article-tags">
                    
                    <a href="/tags/python-advance/">
                        <span class="chip bg-color">python_advance</span>
                    </a>
                    
                    <a href="/tags/%E6%95%B0%E5%80%BC%E5%88%86%E6%9E%90/">
                        <span class="chip bg-color">数值分析</span>
                    </a>
                    
                </div>
                
            </div>
        </div>
        
        
        <div class="article col s12 m6" data-aos="fade-up">
            <div class="article-badge right-badge text-color">
                下一篇&nbsp;<i class="fas fa-chevron-right"></i>
            </div>
            <div class="card">
                <a href="/2023/03/26/mitogenomeassembleannotation/">
                    <div class="card-image">
                        
                        
                        <img src="/medias/plant/0.jpg" class="responsive-img" alt="mitogenomeAssembleAnnotataion">
                        
                        <span class="card-title">mitogenomeAssembleAnnotataion</span>
                    </div>
                </a>
                <div class="card-content article-content">
                    <div class="summary block-with-text">
                        
                            
                        
                    </div>
                    <div class="publish-info">
                            <span class="publish-date">
                                <i class="far fa-clock fa-fw icon-date"></i>2023-03-26
                            </span>
                        <span class="publish-author">
                            
                            <i class="fas fa-user fa-fw"></i>
                            zhangchaofan
                            
                        </span>
                    </div>
                </div>
                
                <div class="card-action article-tags">
                    
                    <a href="/tags/Assemble/">
                        <span class="chip bg-color">Assemble</span>
                    </a>
                    
                    <a href="/tags/mitogenome/">
                        <span class="chip bg-color">mitogenome</span>
                    </a>
                    
                </div>
                
            </div>
        </div>
        
    </div>
</article>

</div>



<!-- 代码块功能依赖 -->
<script type="text/javascript" src="/libs/codeBlock/codeBlockFuction.js"></script>

<!-- 代码语言 -->

<script type="text/javascript" src="/libs/codeBlock/codeLang.js"></script>


<!-- 代码块复制 -->

<script type="text/javascript" src="/libs/codeBlock/codeCopy.js"></script>


<!-- 代码块收缩 -->

<script type="text/javascript" src="/libs/codeBlock/codeShrink.js"></script>


    </div>
    <div id="toc-aside" class="expanded col l3 hide-on-med-and-down">
        <div class="toc-widget card" style="background-color: white;">
            <div class="toc-title"><i class="far fa-list-alt"></i>&nbsp;&nbsp;目录</div>
            <div id="toc-content"></div>
        </div>
    </div>
</div>

<!-- TOC 悬浮按钮. -->

<div id="floating-toc-btn" class="hide-on-med-and-down">
    <a class="btn-floating btn-large bg-color">
        <i class="fas fa-list-ul"></i>
    </a>
</div>


<script src="/libs/tocbot/tocbot.min.js"></script>
<script>
    $(function () {
        tocbot.init({
            tocSelector: '#toc-content',
            contentSelector: '#articleContent',
            headingsOffset: -($(window).height() * 0.4 - 45),
            collapseDepth: Number('0'),
            headingSelector: 'h2, h3, h4, h5'
        });

        // modify the toc link href to support Chinese.
        let i = 0;
        let tocHeading = 'toc-heading-';
        $('#toc-content a').each(function () {
            $(this).attr('href', '#' + tocHeading + (++i));
        });

        // modify the heading title id to support Chinese.
        i = 0;
        $('#articleContent').children('h2, h3, h4, h5').each(function () {
            $(this).attr('id', tocHeading + (++i));
        });

        // Set scroll toc fixed.
        let tocHeight = parseInt($(window).height() * 0.4 - 64);
        let $tocWidget = $('.toc-widget');
        $(window).scroll(function () {
            let scroll = $(window).scrollTop();
            /* add post toc fixed. */
            if (scroll > tocHeight) {
                $tocWidget.addClass('toc-fixed');
            } else {
                $tocWidget.removeClass('toc-fixed');
            }
        });

        
        /* 修复文章卡片 div 的宽度. */
        let fixPostCardWidth = function (srcId, targetId) {
            let srcDiv = $('#' + srcId);
            if (srcDiv.length === 0) {
                return;
            }

            let w = srcDiv.width();
            if (w >= 450) {
                w = w + 21;
            } else if (w >= 350 && w < 450) {
                w = w + 18;
            } else if (w >= 300 && w < 350) {
                w = w + 16;
            } else {
                w = w + 14;
            }
            $('#' + targetId).width(w);
        };

        // 切换TOC目录展开收缩的相关操作.
        const expandedClass = 'expanded';
        let $tocAside = $('#toc-aside');
        let $mainContent = $('#main-content');
        $('#floating-toc-btn .btn-floating').click(function () {
            if ($tocAside.hasClass(expandedClass)) {
                $tocAside.removeClass(expandedClass).hide();
                $mainContent.removeClass('l9');
            } else {
                $tocAside.addClass(expandedClass).show();
                $mainContent.addClass('l9');
            }
            fixPostCardWidth('artDetail', 'prenext-posts');
        });
        
    });
</script>

    

</main>




    <footer class="page-footer bg-color">
    
        <link rel="stylesheet" href="/libs/aplayer/APlayer.min.css">
<style>
    .aplayer .aplayer-lrc p {
        
        display: none;
        
        font-size: 12px;
        font-weight: 700;
        line-height: 16px !important;
    }

    .aplayer .aplayer-lrc p.aplayer-lrc-current {
        
        display: none;
        
        font-size: 15px;
        color: #42b983;
    }

    
    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body {
        left: -66px !important;
    }

    .aplayer.aplayer-fixed.aplayer-narrow .aplayer-body:hover {
        left: 0px !important;
    }

    
</style>
<div class="">
    
    <div class="row">
        <meting-js class="col l8 offset-l2 m10 offset-m1 s12"
                   server="netease"
                   type="playlist"
                   id="2032721327"
                   fixed='true'
                   autoplay='false'
                   theme='#42b983'
                   loop='all'
                   order='random'
                   preload='auto'
                   volume='0.7'
                   list-folded='true'
        >
        </meting-js>
    </div>
</div>

<script src="/libs/aplayer/APlayer.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/meting@2/dist/Meting.min.js"></script>

    
    <div class="container row center-align" style="margin-bottom: 15px !important;">
        <div class="col s12 m8 l8 copy-right">
            Copyright&nbsp;&copy;
            
                <span id="year">2021-2024</span>
            
            <span id="year">2021</span>
            <a href="/about" target="_blank">zhangchaofan</a>
            |&nbsp;Powered by&nbsp;<a href="https://hexo.io/" target="_blank">Hexo</a>
            |&nbsp;Theme&nbsp;<a href="https://github.com/blinkfox/hexo-theme-matery" target="_blank">Matery</a>
            <br>
            
            &nbsp;<i class="fas fa-chart-area"></i>&nbsp;站点总字数:&nbsp;<span
                class="white-color">20.4k</span>&nbsp;字
            
            
            
            
            
            
            <span id="busuanzi_container_site_pv">
                |&nbsp;<i class="far fa-eye"></i>&nbsp;总访问量:&nbsp;<span id="busuanzi_value_site_pv"
                    class="white-color"></span>&nbsp;次
            </span>
            
            
            <span id="busuanzi_container_site_uv">
                |&nbsp;<i class="fas fa-users"></i>&nbsp;总访问人数:&nbsp;<span id="busuanzi_value_site_uv"
                    class="white-color"></span>&nbsp;人
            </span>
            
            <br>
            
            <span id="sitetime">载入运行时间...</span>
            <script>
                function siteTime() {
                    var seconds = 1000;
                    var minutes = seconds * 60;
                    var hours = minutes * 60;
                    var days = hours * 24;
                    var years = days * 365;
                    var today = new Date();
                    var startYear = "2021";
                    var startMonth = "3";
                    var startDate = "09";
                    var startHour = "0";
                    var startMinute = "0";
                    var startSecond = "0";
                    var todayYear = today.getFullYear();
                    var todayMonth = today.getMonth() + 1;
                    var todayDate = today.getDate();
                    var todayHour = today.getHours();
                    var todayMinute = today.getMinutes();
                    var todaySecond = today.getSeconds();
                    var t1 = Date.UTC(startYear, startMonth, startDate, startHour, startMinute, startSecond);
                    var t2 = Date.UTC(todayYear, todayMonth, todayDate, todayHour, todayMinute, todaySecond);
                    var diff = t2 - t1;
                    var diffYears = Math.floor(diff / years);
                    var diffDays = Math.floor((diff / days) - diffYears * 365);
                    var diffHours = Math.floor((diff - (diffYears * 365 + diffDays) * days) / hours);
                    var diffMinutes = Math.floor((diff - (diffYears * 365 + diffDays) * days - diffHours * hours) /
                        minutes);
                    var diffSeconds = Math.floor((diff - (diffYears * 365 + diffDays) * days - diffHours * hours -
                        diffMinutes * minutes) / seconds);
                    if (startYear == todayYear) {
                        document.getElementById("year").innerHTML = todayYear;
                        document.getElementById("sitetime").innerHTML = "本站已安全运行 " + diffDays + " 天 " + diffHours +
                            " 小时 " + diffMinutes + " 分钟 " + diffSeconds + " 秒";
                    } else {
                        document.getElementById("year").innerHTML = startYear + " - " + todayYear;
                        document.getElementById("sitetime").innerHTML = "本站已安全运行 " + diffYears + " 年 " + diffDays +
                            " 天 " + diffHours + " 小时 " + diffMinutes + " 分钟 " + diffSeconds + " 秒";
                    }
                }
                setInterval(siteTime, 1000);
            </script>
            
            <br>
            
        </div>
        <div class="col s12 m4 l4 social-link social-statis">
    <a href="https://github.com/chaofan520" class="tooltipped" target="_blank" data-tooltip="访问我的GitHub" data-position="top" data-delay="50">
        <i class="fab fa-github"></i>
    </a>



    <a href="mailto:2439555626@qq.com" class="tooltipped" target="_blank" data-tooltip="邮件联系我" data-position="top" data-delay="50">
        <i class="fas fa-envelope-open"></i>
    </a>







    <a href="tencent://AddContact/?fromId=50&fromSubId=1&subcmd=all&uin=2439555626" class="tooltipped" target="_blank" data-tooltip="QQ联系我: 2439555626" data-position="top" data-delay="50">
        <i class="fab fa-qq"></i>
    </a>














</div>
    </div>
</footer>

<div class="progress-bar"></div>


    <!-- 搜索遮罩框 -->
<div id="searchModal" class="modal">
    <div class="modal-content">
        <div class="search-header">
            <span class="title"><i class="fas fa-search"></i>&nbsp;&nbsp;搜索</span>
            <input type="search" id="searchInput" name="s" placeholder="请输入搜索的关键字"
                   class="search-input">
        </div>
        <div id="searchResult"></div>
    </div>
</div>

<script type="text/javascript">
$(function () {
    var searchFunc = function (path, search_id, content_id) {
        'use strict';
        $.ajax({
            url: path,
            dataType: "xml",
            success: function (xmlResponse) {
                // get the contents from search data
                var datas = $("entry", xmlResponse).map(function () {
                    return {
                        title: $("title", this).text(),
                        content: $("content", this).text(),
                        url: $("url", this).text()
                    };
                }).get();
                var $input = document.getElementById(search_id);
                var $resultContent = document.getElementById(content_id);
                $input.addEventListener('input', function () {
                    var str = '<ul class=\"search-result-list\">';
                    var keywords = this.value.trim().toLowerCase().split(/[\s\-]+/);
                    $resultContent.innerHTML = "";
                    if (this.value.trim().length <= 0) {
                        return;
                    }
                    // perform local searching
                    datas.forEach(function (data) {
                        var isMatch = true;
                        var data_title = data.title.trim().toLowerCase();
                        var data_content = data.content.trim().replace(/<[^>]+>/g, "").toLowerCase();
                        var data_url = data.url;
                        data_url = data_url.indexOf('/') === 0 ? data.url : '/' + data_url;
                        var index_title = -1;
                        var index_content = -1;
                        var first_occur = -1;
                        // only match artiles with not empty titles and contents
                        if (data_title !== '' && data_content !== '') {
                            keywords.forEach(function (keyword, i) {
                                index_title = data_title.indexOf(keyword);
                                index_content = data_content.indexOf(keyword);
                                if (index_title < 0 && index_content < 0) {
                                    isMatch = false;
                                } else {
                                    if (index_content < 0) {
                                        index_content = 0;
                                    }
                                    if (i === 0) {
                                        first_occur = index_content;
                                    }
                                }
                            });
                        }
                        // show search results
                        if (isMatch) {
                            str += "<li><a href='" + data_url + "' class='search-result-title'>" + data_title + "</a>";
                            var content = data.content.trim().replace(/<[^>]+>/g, "");
                            if (first_occur >= 0) {
                                // cut out 100 characters
                                var start = first_occur - 20;
                                var end = first_occur + 80;
                                if (start < 0) {
                                    start = 0;
                                }
                                if (start === 0) {
                                    end = 100;
                                }
                                if (end > content.length) {
                                    end = content.length;
                                }
                                var match_content = content.substr(start, end);
                                // highlight all keywords
                                keywords.forEach(function (keyword) {
                                    var regS = new RegExp(keyword, "gi");
                                    match_content = match_content.replace(regS, "<em class=\"search-keyword\">" + keyword + "</em>");
                                });

                                str += "<p class=\"search-result\">" + match_content + "...</p>"
                            }
                            str += "</li>";
                        }
                    });
                    str += "</ul>";
                    $resultContent.innerHTML = str;
                });
            }
        });
    };

    searchFunc('/search.xml', 'searchInput', 'searchResult');
});
</script>

    <!-- 回到顶部按钮 -->
<div id="backTop" class="top-scroll">
    <a class="btn-floating btn-large waves-effect waves-light" href="#!">
        <i class="fas fa-arrow-up"></i>
    </a>
</div>


    <script src="/libs/materialize/materialize.min.js"></script>
    <script src="/libs/masonry/masonry.pkgd.min.js"></script>
    <script src="/libs/aos/aos.js"></script>
    <script src="/libs/scrollprogress/scrollProgress.min.js"></script>
    <script src="/libs/lightGallery/js/lightgallery-all.min.js"></script>
    <script src="/js/matery.js"></script>

    <!-- Baidu Analytics -->

    <!-- Baidu Push -->

<script>
    (function () {
        var bp = document.createElement('script');
        var curProtocol = window.location.protocol.split(':')[0];
        if (curProtocol === 'https') {
            bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';
        } else {
            bp.src = 'http://push.zhanzhang.baidu.com/push.js';
        }
        var s = document.getElementsByTagName("script")[0];
        s.parentNode.insertBefore(bp, s);
    })();
</script>

    
    <script src="/libs/others/clicklove.js" async="async"></script>
    
	
    <script async src="/libs/others/busuanzi.pure.mini.js"></script>
    

    

    

    <!--腾讯兔小巢-->
    <!-- 
     -->

    
    
    <script type="text/javascript" size="150" alpha='0.6'
        zIndex="-1" src="/libs/background/ribbon.min.js" async="async"></script>
    

    
    <script type="text/javascript" src="/libs/background/ribbon-dynamic.js" async="async"></script>
    

    
    <script src="/libs/instantpage/instantpage.js" type="module"></script>
     
    <script src="/js/cursor.js"></script>
    <script src="/js/FunnyTitle.js"></script>
    <script src="https://cdn.jsdelivr.net/gh/stevenjoezhang/live2d-widget@latest/autoload.js"></script> 
    <!-- 动态线条背景
    <script type="text/javascript"
    color="122 103 238" opacity='0.7' zIndex="-2" count="200" src="//cdn.bootcss.com/canvas-nest.js/1.0.0/canvas-nest.min.js">
    </script> -->
    
    <script type="text/javascript">
    //只在桌面版网页启用特效
    var windowWidth = $(window).width();
    if (windowWidth > 768) {
        document.write('<script type="text/javascript" src="/js/sakura.js"><\/script>');
    }
    </script>
    
</body>

</html>
